File: backend/app/routers/sync.py
from datetime import UTC, datetime

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app import models, schemas
from app.auth import get_current_user
from app.database import get_db

router = APIRouter()


def _upsert_recipe(db: Session, user: models.User, item: schemas.RecipeSyncItem) -> bool:
    recipe = (
        db.query(models.Recipe)
        .filter(models.Recipe.public_id == item.id, models.Recipe.user_id == user.id)
        .first()
    )
    data = item.model_dump(exclude={"id", "created_at"})
    if recipe is None:
        db.add(
            models.Recipe(public_id=item.id, user_id=user.id, created_at=item.created_at, **data)
        )
        return True
    for field, value in data.items():
        setattr(recipe, field, value)
    return False


def _upsert_pour_step(
    db: Session, user: models.User, item: schemas.PourStepSyncItem
) -> bool | None:
    recipe = (
        db.query(models.Recipe)
        .filter(models.Recipe.public_id == item.recipe_id, models.Recipe.user_id == user.id)
        .first()
    )
    if recipe is None:
        return None
    step = (
        db.query(models.PourStep)
        .join(models.Recipe, models.PourStep.recipe_id == models.Recipe.id)
        .filter(models.PourStep.public_id == item.id, models.Recipe.user_id == user.id)
        .first()
    )
    data = item.model_dump(exclude={"id", "recipe_id"})
    if step is None:
        db.add(models.PourStep(public_id=item.id, recipe_id=recipe.id, **data))
        return True
    for field, value in data.items():
        setattr(step, field, value)
    return False


def _upsert_brew_log(db: Session, user: models.User, item: schemas.BrewLogSyncItem) -> bool | None:
    recipe = (
        db.query(models.Recipe)
        .filter(models.Recipe.public_id == item.recipe_id, models.Recipe.user_id == user.id)
        .first()
    )
    if recipe is None:
        return None
    log = (
        db.query(models.BrewLog)
        .join(models.Recipe, models.BrewLog.recipe_id == models.Recipe.id)
        .filter(models.BrewLog.public_id == item.id, models.Recipe.user_id == user.id)
        .first()
    )
    data = item.model_dump(exclude={"id", "recipe_id"})
    if log is None:
        db.add(models.BrewLog(public_id=item.id, recipe_id=recipe.id, **data))
        return True
    for field, value in data.items():
        setattr(log, field, value)
    return False


def _soft_delete_recipe(db: Session, user: models.User, public_id: str) -> bool:
    recipe = (
        db.query(models.Recipe)
        .filter(models.Recipe.public_id == public_id, models.Recipe.user_id == user.id)
        .first()
    )
    if recipe is None:
        return False
    now = datetime.now(UTC)
    setattr(recipe, "deleted_at", now)  # noqa: B010 -- ty misreads Column[] attrs on direct assign
    db.query(models.PourStep).filter(models.PourStep.recipe_id == recipe.id).update(
        {"deleted_at": now}
    )
    db.query(models.BrewLog).filter(models.BrewLog.recipe_id == recipe.id).update(
        {"deleted_at": now}
    )
    return True


def _soft_delete_pour_step(db: Session, user: models.User, public_id: str) -> bool:
    step = (
        db.query(models.PourStep)
        .join(models.Recipe, models.PourStep.recipe_id == models.Recipe.id)
        .filter(models.PourStep.public_id == public_id, models.Recipe.user_id == user.id)
        .first()
    )
    if step is None:
        return False
    setattr(step, "deleted_at", datetime.now(UTC))  # noqa: B010 -- see _soft_delete_recipe
    return True


def _soft_delete_brew_log(db: Session, user: models.User, public_id: str) -> bool:
    log = (
        db.query(models.BrewLog)
        .join(models.Recipe, models.BrewLog.recipe_id == models.Recipe.id)
        .filter(models.BrewLog.public_id == public_id, models.Recipe.user_id == user.id)
        .first()
    )
    if log is None:
        return False
    setattr(log, "deleted_at", datetime.now(UTC))  # noqa: B010 -- see _soft_delete_recipe
    return True


@router.post("/push", response_model=schemas.SyncPushResponse)
def push(
    payload: schemas.SyncPushRequest,
    current_user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    recipes_upserted = sum(_upsert_recipe(db, current_user, item) for item in payload.recipes)
    db.flush()  # newly-created recipes must be visible to pour_step/brew_log parent lookups below
    pour_steps_upserted = sum(
        bool(_upsert_pour_step(db, current_user, item)) for item in payload.pour_steps
    )
    brew_logs_upserted = sum(
        bool(_upsert_brew_log(db, current_user, item)) for item in payload.brew_logs
    )
    pour_steps_deleted = sum(
        _soft_delete_pour_step(db, current_user, public_id)
        for public_id in payload.pour_steps_deleted
    )
    brew_logs_deleted = sum(
        _soft_delete_brew_log(db, current_user, public_id)
        for public_id in payload.brew_logs_deleted
    )
    recipes_deleted = sum(
        _soft_delete_recipe(db, current_user, public_id) for public_id in payload.recipes_deleted
    )
    db.commit()
    return schemas.SyncPushResponse(
        recipes_upserted=recipes_upserted,
        recipes_deleted=recipes_deleted,
        pour_steps_upserted=pour_steps_upserted,
        pour_steps_deleted=pour_steps_deleted,
        brew_logs_upserted=brew_logs_upserted,
        brew_logs_deleted=brew_logs_deleted,
    )


@router.get("/pull", response_model=schemas.SyncPullResponse)
def pull(
    current_user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    recipes = (
        db.query(models.Recipe)
        .filter(models.Recipe.user_id == current_user.id, models.Recipe.deleted_at.is_(None))
        .all()
    )
    pour_steps = (
        db.query(models.PourStep)
        .join(models.Recipe, models.PourStep.recipe_id == models.Recipe.id)
        .filter(
            models.Recipe.user_id == current_user.id,
            models.Recipe.deleted_at.is_(None),
            models.PourStep.deleted_at.is_(None),
        )
        .all()
    )
    brew_logs = (
        db.query(models.BrewLog)
        .join(models.Recipe, models.BrewLog.recipe_id == models.Recipe.id)
        .filter(
            models.Recipe.user_id == current_user.id,
            models.Recipe.deleted_at.is_(None),
            models.BrewLog.deleted_at.is_(None),
        )
        .all()
    )
    return schemas.SyncPullResponse(
        recipes=[
            schemas.RecipeSyncItem.model_validate(
                {
                    "id": r.public_id,
                    "name": r.name,
                    "bean_origin": r.bean_origin,
                    "dose_g": r.dose_g,
                    "water_ml": r.water_ml,
                    "water_temp_c": r.water_temp_c,
                    "grind_size": r.grind_size,
                    "total_time_sec": r.total_time_sec,
                    "notes": r.notes,
                    "created_at": r.created_at,
                    "updated_at": r.updated_at,
                }
            )
            for r in recipes
        ],
        pour_steps=[
            schemas.PourStepSyncItem.model_validate(
                {
                    "id": s.public_id,
                    "recipe_id": s.recipe.public_id,
                    "step_order": s.step_order,
                    "target_time_sec": s.target_time_sec,
                    "cumulative_water_ml": s.cumulative_water_ml,
                    "notes": s.notes,
                }
            )
            for s in pour_steps
        ],
        brew_logs=[
            schemas.BrewLogSyncItem.model_validate(
                {
                    "id": log.public_id,
                    "recipe_id": log.recipe.public_id,
                    "brewed_at": log.brewed_at,
                    "rating": log.rating,
                    "notes": log.notes,
                    "created_at": log.created_at,
                }
            )
            for log in brew_logs
        ],
    )

File: worker/src/routes/sync.ts
import { and, eq, isNull } from "drizzle-orm";
import { drizzle } from "drizzle-orm/d1";
import { Hono } from "hono";
import { z } from "zod";
import { createAuth } from "../auth";
import * as schema from "../db/schema";
import { isSafePurchaseUrl } from "../lib/url";

type Database = ReturnType<typeof drizzle<typeof schema>>;

// Field lists mirror the frontend's IndexedDB-backed Recipe/PourStep/BrewLog types
// (and the now-retired backend/app/schemas.py Sync* models) exactly, so the client
// can push/pull without reshaping. `id` is always the resource's client-generated
// UUID (stored server-side as `publicId`, with no server-side default — unlike
// better-auth's own ids, the client always supplies it). A PourStep/BrewLog's
// `recipe_id` is always the parent Recipe's `id` (its `publicId`), never its
// internal integer id, per .claude/skills/secure-resource-access/SKILL.md.

const recipeSyncItemSchema = z.object({
  id: z.string(),
  name: z.string(),
  bean_origin: z.string().nullable().optional(),
  dose_g: z.number().positive(),
  water_ml: z.number().positive(),
  water_temp_c: z.number().positive(),
  grind_size: z.string().nullable().optional(),
  total_time_sec: z.number().int().nonnegative().nullable().optional(),
  notes: z.string().nullable().optional(),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

const pourStepSyncItemSchema = z.object({
  id: z.string(),
  recipe_id: z.string(),
  step_order: z.number().int(),
  target_time_sec: z.number().int().nonnegative(),
  cumulative_water_ml: z.number().positive(),
  notes: z.string().nullable().optional(),
});

const brewLogSyncItemSchema = z.object({
  id: z.string(),
  recipe_id: z.string(),
  brewed_at: z.coerce.date(),
  rating: z.number().int().min(1).max(5),
  notes: z.string().nullable().optional(),
  created_at: z.coerce.date(),
});

// purchase_url is validated against isSafePurchaseUrl (http/https only, no
// credentials, no loopback/private/link-local host) as defense-in-depth --
// see worker/src/lib/url.ts for the full rationale (this app never fetches
// the URL server-side, so there is no live SSRF surface today).
const beanSyncItemSchema = z.object({
  id: z.string(),
  name: z.string().min(1).max(200),
  origin: z.string().max(200).nullable().optional(),
  roaster: z.string().max(200).nullable().optional(),
  roast_level: z.string().max(100).nullable().optional(),
  roast_date: z.coerce.date().nullable().optional(),
  purchase_url: z
    .string()
    .max(2048)
    .refine(isSafePurchaseUrl, { message: "purchase_url must be a safe http(s) URL" })
    .nullable()
    .optional(),
  notes: z.string().nullable().optional(),
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

const syncPushRequestSchema = z.object({
  recipes: z.array(recipeSyncItemSchema).default([]),
  recipes_deleted: z.array(z.string()).default([]),
  pour_steps: z.array(pourStepSyncItemSchema).default([]),
  pour_steps_deleted: z.array(z.string()).default([]),
  brew_logs: z.array(brewLogSyncItemSchema).default([]),
  brew_logs_deleted: z.array(z.string()).default([]),
  beans: z.array(beanSyncItemSchema).default([]),
  beans_deleted: z.array(z.string()).default([]),
});

type RecipeSyncItem = z.infer<typeof recipeSyncItemSchema>;
type PourStepSyncItem = z.infer<typeof pourStepSyncItemSchema>;
type BrewLogSyncItem = z.infer<typeof brewLogSyncItemSchema>;
type BeanSyncItem = z.infer<typeof beanSyncItemSchema>;

async function upsertRecipe(db: Database, userId: string, item: RecipeSyncItem): Promise<boolean> {
  const existing = await db.query.recipe.findFirst({
    where: and(eq(schema.recipe.publicId, item.id), eq(schema.recipe.userId, userId)),
  });
  const data = {
    name: item.name,
    beanOrigin: item.bean_origin ?? null,
    doseG: item.dose_g,
    waterMl: item.water_ml,
    waterTempC: item.water_temp_c,
    grindSize: item.grind_size ?? null,
    totalTimeSec: item.total_time_sec ?? null,
    notes: item.notes ?? null,
    updatedAt: item.updated_at,
  };
  if (!existing) {
    await db.insert(schema.recipe).values({
      publicId: item.id,
      userId,
      createdAt: item.created_at,
      ...data,
    });
    return true;
  }
  await db.update(schema.recipe).set(data).where(eq(schema.recipe.id, existing.id));
  return false;
}

async function upsertPourStep(
  db: Database,
  userId: string,
  item: PourStepSyncItem,
): Promise<boolean | null> {
  const parent = await db.query.recipe.findFirst({
    where: and(eq(schema.recipe.publicId, item.recipe_id), eq(schema.recipe.userId, userId)),
  });
  if (!parent) return null;
  const existing = await db.query.pourStep.findFirst({
    where: and(eq(schema.pourStep.publicId, item.id), eq(schema.pourStep.recipeId, parent.id)),
  });
  const data = {
    stepOrder: item.step_order,
    targetTimeSec: item.target_time_sec,
    cumulativeWaterMl: item.cumulative_water_ml,
    notes: item.notes ?? null,
  };
  if (!existing) {
    await db.insert(schema.pourStep).values({ publicId: item.id, recipeId: parent.id, ...data });
    return true;
  }
  await db.update(schema.pourStep).set(data).where(eq(schema.pourStep.id, existing.id));
  return false;
}

async function upsertBrewLog(
  db: Database,
  userId: string,
  item: BrewLogSyncItem,
): Promise<boolean | null> {
  const parent = await db.query.recipe.findFirst({
    where: and(eq(schema.recipe.publicId, item.recipe_id), eq(schema.recipe.userId, userId)),
  });
  if (!parent) return null;
  const existing = await db.query.brewLog.findFirst({
    where: and(eq(schema.brewLog.publicId, item.id), eq(schema.brewLog.recipeId, parent.id)),
  });
  const data = {
    brewedAt: item.brewed_at,
    rating: item.rating,
    notes: item.notes ?? null,
  };
  if (!existing) {
    await db.insert(schema.brewLog).values({
      publicId: item.id,
      recipeId: parent.id,
      createdAt: item.created_at,
      ...data,
    });
    return true;
  }
  await db.update(schema.brewLog).set(data).where(eq(schema.brewLog.id, existing.id));
  return false;
}

async function upsertBean(db: Database, userId: string, item: BeanSyncItem): Promise<boolean> {
  const existing = await db.query.bean.findFirst({
    where: and(eq(schema.bean.publicId, item.id), eq(schema.bean.userId, userId)),
  });
  const data = {
    name: item.name,
    origin: item.origin ?? null,
    roaster: item.roaster ?? null,
    roastLevel: item.roast_level ?? null,
    roastDate: item.roast_date ?? null,
    purchaseUrl: item.purchase_url ?? null,
    notes: item.notes ?? null,
    updatedAt: item.updated_at,
  };
  if (!existing) {
    await db.insert(schema.bean).values({
      publicId: item.id,
      userId,
      createdAt: item.created_at,
      ...data,
    });
    return true;
  }
  await db.update(schema.bean).set(data).where(eq(schema.bean.id, existing.id));
  return false;
}

async function softDeleteBean(db: Database, userId: string, publicId: string): Promise<boolean> {
  const existing = await db.query.bean.findFirst({
    where: and(eq(schema.bean.publicId, publicId), eq(schema.bean.userId, userId)),
  });
  if (!existing) return false;
  await db
    .update(schema.bean)
    .set({ deletedAt: new Date() })
    .where(eq(schema.bean.id, existing.id));
  return true;
}

async function softDeleteRecipe(db: Database, userId: string, publicId: string): Promise<boolean> {
  const existing = await db.query.recipe.findFirst({
    where: and(eq(schema.recipe.publicId, publicId), eq(schema.recipe.userId, userId)),
  });
  if (!existing) return false;
  const now = new Date();
  await db.update(schema.recipe).set({ deletedAt: now }).where(eq(schema.recipe.id, existing.id));
  await db
    .update(schema.pourStep)
    .set({ deletedAt: now })
    .where(eq(schema.pourStep.recipeId, existing.id));
  await db
    .update(schema.brewLog)
    .set({ deletedAt: now })
    .where(eq(schema.brewLog.recipeId, existing.id));
  return true;
}

async function softDeletePourStep(
  db: Database,
  userId: string,
  publicId: string,
): Promise<boolean> {
  const existing = await db
    .select({ id: schema.pourStep.id })
    .from(schema.pourStep)
    .innerJoin(schema.recipe, eq(schema.pourStep.recipeId, schema.recipe.id))
    .where(and(eq(schema.pourStep.publicId, publicId), eq(schema.recipe.userId, userId)))
    .get();
  if (!existing) return false;
  await db
    .update(schema.pourStep)
    .set({ deletedAt: new Date() })
    .where(eq(schema.pourStep.id, existing.id));
  return true;
}

async function softDeleteBrewLog(db: Database, userId: string, publicId: string): Promise<boolean> {
  const existing = await db
    .select({ id: schema.brewLog.id })
    .from(schema.brewLog)
    .innerJoin(schema.recipe, eq(schema.brewLog.recipeId, schema.recipe.id))
    .where(and(eq(schema.brewLog.publicId, publicId), eq(schema.recipe.userId, userId)))
    .get();
  if (!existing) return false;
  await db
    .update(schema.brewLog)
    .set({ deletedAt: new Date() })
    .where(eq(schema.brewLog.id, existing.id));
  return true;
}

export const syncApp = new Hono<{ Bindings: Env; Variables: { userId: string } }>();

syncApp.use("*", async (c, next) => {
  const session = await createAuth(c.env).api.getSession({ headers: c.req.raw.headers });
  if (!session) return c.json({ message: "Not authenticated" }, 401);
  c.set("userId", session.user.id);
  await next();
});

syncApp.post("/push", async (c) => {
  const payload = syncPushRequestSchema.parse(await c.req.json());
  const userId = c.get("userId");
  const db = drizzle(c.env.DB, { schema });

  let beansUpserted = 0;
  for (const item of payload.beans) {
    if (await upsertBean(db, userId, item)) beansUpserted++;
  }
  let beansDeleted = 0;
  for (const publicId of payload.beans_deleted) {
    if (await softDeleteBean(db, userId, publicId)) beansDeleted++;
  }
  let recipesUpserted = 0;
  for (const item of payload.recipes) {
    if (await upsertRecipe(db, userId, item)) recipesUpserted++;
  }
  let pourStepsUpserted = 0;
  for (const item of payload.pour_steps) {
    if (await upsertPourStep(db, userId, item)) pourStepsUpserted++;
  }
  let brewLogsUpserted = 0;
  for (const item of payload.brew_logs) {
    if (await upsertBrewLog(db, userId, item)) brewLogsUpserted++;
  }
  let pourStepsDeleted = 0;
  for (const publicId of payload.pour_steps_deleted) {
    if (await softDeletePourStep(db, userId, publicId)) pourStepsDeleted++;
  }
  let brewLogsDeleted = 0;
  for (const publicId of payload.brew_logs_deleted) {
    if (await softDeleteBrewLog(db, userId, publicId)) brewLogsDeleted++;
  }
  let recipesDeleted = 0;
  for (const publicId of payload.recipes_deleted) {
    if (await softDeleteRecipe(db, userId, publicId)) recipesDeleted++;
  }

  return c.json({
    recipes_upserted: recipesUpserted,
    recipes_deleted: recipesDeleted,
    pour_steps_upserted: pourStepsUpserted,
    pour_steps_deleted: pourStepsDeleted,
    brew_logs_upserted: brewLogsUpserted,
    brew_logs_deleted: brewLogsDeleted,
    beans_upserted: beansUpserted,
    beans_deleted: beansDeleted,
  });
});

syncApp.get("/pull", async (c) => {
  const userId = c.get("userId");
  const db = drizzle(c.env.DB, { schema });

  const beans = await db.query.bean.findMany({
    where: and(eq(schema.bean.userId, userId), isNull(schema.bean.deletedAt)),
  });
  const recipes = await db.query.recipe.findMany({
    where: and(eq(schema.recipe.userId, userId), isNull(schema.recipe.deletedAt)),
  });
  const pourSteps = await db
    .select({
      id: schema.pourStep.publicId,
      recipe_id: schema.recipe.publicId,
      step_order: schema.pourStep.stepOrder,
      target_time_sec: schema.pourStep.targetTimeSec,
      cumulative_water_ml: schema.pourStep.cumulativeWaterMl,
      notes: schema.pourStep.notes,
    })
    .from(schema.pourStep)
    .innerJoin(schema.recipe, eq(schema.pourStep.recipeId, schema.recipe.id))
    .where(
      and(
        eq(schema.recipe.userId, userId),
        isNull(schema.recipe.deletedAt),
        isNull(schema.pourStep.deletedAt),
      ),
    );
  const brewLogs = await db
    .select({
      id: schema.brewLog.publicId,
      recipe_id: schema.recipe.publicId,
      brewed_at: schema.brewLog.brewedAt,
      rating: schema.brewLog.rating,
      notes: schema.brewLog.notes,
      created_at: schema.brewLog.createdAt,
    })
    .from(schema.brewLog)
    .innerJoin(schema.recipe, eq(schema.brewLog.recipeId, schema.recipe.id))
    .where(
      and(
        eq(schema.recipe.userId, userId),
        isNull(schema.recipe.deletedAt),
        isNull(schema.brewLog.deletedAt),
      ),
    );

  return c.json({
    recipes: recipes.map((r) => ({
      id: r.publicId,
      name: r.name,
      bean_origin: r.beanOrigin,
      dose_g: r.doseG,
      water_ml: r.waterMl,
      water_temp_c: r.waterTempC,
      grind_size: r.grindSize,
      total_time_sec: r.totalTimeSec,
      notes: r.notes,
      created_at: r.createdAt,
      updated_at: r.updatedAt,
    })),
    pour_steps: pourSteps,
    brew_logs: brewLogs,
    beans: beans.map((b) => ({
      id: b.publicId,
      name: b.name,
      origin: b.origin,
      roaster: b.roaster,
      roast_level: b.roastLevel,
      roast_date: b.roastDate,
      purchase_url: b.purchaseUrl,
      notes: b.notes,
      created_at: b.createdAt,
      updated_at: b.updatedAt,
    })),
  });
});
